### ISSUE TITLE
Edge TPU Compiler Failed to Compile

### ISSUE THREAD

**Issue by rgbarishian** [tensorflow/tensorflow#41069]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.4 LTS
- TensorFlow installed from (source or binary): anaconda
- TensorFlow version (or github SHA if from source): 2.2.0
- Edge TPU Compiler version 2.1.302470888

**Command used to run the converter or code if you’re using the Python API**

```
edgetpu_compiler quantized.tflite
```

**The output from the converter invocation**

```
Edge TPU Compiler version 2.1.302470888
ERROR: :129 std::abs(input_product_scale - bias_scale) <= 1e-6 * std::min(input_product_scale, bias_scale) was not true.
ERROR: Node number 40 (FULLY_CONNECTED) failed to prepare.
```

**Failure details**
I have used the same quantizing and converting to tflite code before and it works fine on a different structured model. I use it to convert .h5 keras models to tflite models. It works if I create a model based off MobileNet with additional dense layers
```
base_model = MobileNet(weights='imagenet', include_top=False, input_shape=(200,200,3), dropout=.2)
x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(1024, activation='relu')(x)
x = Dense(1024, activation='relu')(x)
x = Dense(512,activation='relu')(x)
types = Dense(20,activation='softmax')(x)
model = Model(inputs=base_model.input,outputs=types)
```

But if I try to convert model with 2 outputs like this, I get the aformentioned error when compiling.
```
base_model = MobileNet(weights='imagenet', include_top=False, input_shape=(200,200,3), dropout=.2)
x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(1024, activation='relu')(x)
x = Dense(1024, activation='relu')(x)
x = Dense(512,activation='relu')(x)
types = Dense(20,activation='softmax')(x)

y = base_model.output
y = GlobalAveragePooling2D()(y)
y = Dense(512, activation='relu')(y)
y = Dense(1024, activation='relu')(y)
y = Dense(512, activation='relu')(y)
values = Dense(3, activation='sigmoid')(y)

model = Model(inputs=base_model.input, outputs=[types,values])
```

I can get the same error if I attempt to load the quantized model.
```
interpreter = tf.lite.Interpreter(model_path='path/to/model/quantized.tflite')
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

print("shape:", output_details[0]['shape'])
print("shape:", output_details[1]['shape'])
```
Output is as expected:
```
shape: [ 1 20]
shape: [1 3]
```

If I then
```
#reshape for input of batch size 32
interpreter.resize_tensor_input(input_details[0]['index'], (32, 200, 200, 3))
interpreter.resize_tensor_input(output_details[0]['index'], (32, 20))
#interpreter.resize_tensor_input(output_details[1]['index'], (32, 3))
interpreter.allocate_tensors()
```
**Output:**
```
RuntimeError                              Traceback (most recent call last)
 in 
      3 interpreter.resize_tensor_input(output_details[0]['index'], (32, 20))
      4 # interpreter.resize_tensor_input(output_details[1]['index'], (32, 3))
----> 5 interpreter.allocate_tensors()
      6 

~/Software/anaconda3/envs/Tensorflow2/lib/python3.7/site-packages/tensorflow_core/lite/python/interpreter.py in allocate_tensors(self)
    245   def allocate_tensors(self):
    246     self._ensure_safe()
--> 247     return self._interpreter.AllocateTensors()
    248 
    249   def _safe_to_run(self):

~/Software/anaconda3/envs/Tensorflow2/lib/python3.7/site-packages/tensorflow_core/lite/python/interpreter_wrapper/tensorflow_wrap_interpreter_wrapper.py in AllocateTensors(self)
    108 
    109     def AllocateTensors(self):
--> 110         return _tensorflow_wrap_interpreter_wrapper.InterpreterWrapper_AllocateTensors(self)
    111 
    112     def Invoke(self):

RuntimeError: tensorflow/lite/kernels/kernel_util.cc:106 std::abs(input_product_scale - bias_scale) <= 1e-6 * std::min(input_product_scale, bias_scale) was not true.Node number 40 (FULLY_CONNECTED) failed to prepare.
```
Again, this error does not occur when the model has one output.

**Comment by amahendrakar** [comment]
@rgbarishian,
On running the code, I was unable to convert the MobileNet with dense layers model. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/cf8b652ae402bd122591dcb310c3d760/41069.ipynb).

Could you please provide the complete code to convert the model? Also, please take a look at [this](https://github.com/tensorflow/tensorflow/issues/32496) similar issue and let us know if it helps. Thanks!

**Comment by rgbarishian** [comment]
Here is my conversion code. It is based on this tutorial: https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_classification_ptq_tf1.ipynb
Like I mentioned before, this conversion code works fine if the network has one output, it only fails the compiling after this conversion if it has two outputs
```
import os
import tensorflow as tf
import numpy as np
from PIL import Image
from tensorflow.keras.applications.mobilenet import preprocess_input
import pathlib

model = tf.keras.models.load_model('path/to/file.h5')
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]

os.chdir('/directory/of/image/directories')#Where image directories are
directory = os.listdir()
directory

def representative_dataset_gen():
    for i in directory:
        count = 0
        os.chdir(i)
        files = os.listdir()
        print(i)
        for j in files:
            if count<600:
                img = Image.open(j)
                array = np.asarray(img, dtype=np.float32)
                array = preprocess_input(array)
                count=count+1
                yield[np.expand_dims(array, axis=0)]
            else:
                break      
        os.chdir('../')

converter.representative_dataset = representative_dataset_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.int8  # or tf.uint8
converter.inference_output_type = tf.int8  # or tf.uint8
tflite_quant_model = converter.convert()

#Saving file
tflite_model_dir = pathlib.Path('save/path/')
tflite_quant_model_file = tflite_model_dir/'quantized.tflite'
tflite_quant_model_file.write_bytes(tflite_quant_model)
```

**Comment by rgbarishian** [comment]
I think I figured out what caused the problem, but I don't know why. When transfer learning, I would freeze one output branch and set the loss to 0. If I compiled with some layers never trained, the compilation fails. Once I train the entire network such that all layers have a loss, it successfully compiles.
Here is the compilation and loss code:
```
def null_loss(y_true, y_pred):
        return K.zeros_like(y_true)

losses = {
        'Output1': 'categorical_crossentropy',
        'Output2': null_loss
        }
model.compile(optimizer=optimizer, loss=losses, metrics={Output1':'accuracy'})
```

**Comment by karimnosseir** [comment]
Sorry, i am confused by different replies.
Are you having problem to convert when you set 2 outputs or compile the model, or something else ?

Can you please explain what is the issue you are reporting here exactly ?

Thanks

**Comment by rgbarishian** [comment]
That is correct. I am having trouble using the EdgeTPU Compiler with a model with two outputs.
When I have a network with one output, I can compile fine. The compiler gave the error in my original post when my two output model was not entirely trained (One branch had its weights frozen and had the output had a `null_loss` (1keras.backend.zeros_like()`). I think the problem was caused because some layers were never trained and possibly the loss was 0. When I finally finished training, and then ran the compiler, it successfully compiled.

Interestingly, at the end of my training, I freeze the base of the network and the first branch that was originally training. I also set that output's loss to `null_loss` to prevent it from backpropogating. This is when I train the second branch. Once the second branch trains for at least one epoch, I can convert to tflite and compile successfully.
